bgbb/sql/sql_utils.py

```diff
@@ -0,0 +1,219 @@
+    FROM cid_day C
+    WHERE
+      C.submission_date_s3 >= '{ho_start_date_str}'
+      AND C.submission_date_s3 <= '{ho_last_date_str}'
+    GROUP BY 1
+)
+
+, rec_freq as (
+    SELECT
+        C.client_id
+        , datediff(C.Max_day, Min_day) Recency
+        , X - 1 Frequency
+        , datediff('{ho_start_date}', Min_day) - 1 T
+        , C.Max_day
+        , C.Min_day
+    FROM cid_model C
+)
+
+, rec_freq_holdout as (
+  SELECT R.*
+        , coalesce(H.N_holdout, 0) N_holdout
+  FROM rec_freq R
+  LEFT JOIN cid_holdout H
+    ON R.client_id = H.client_id
+)
+
+SELECT * FROM {qname}
+'''
+
+
+def mk_rec_freq_q(q=None, holdout=False, model_start_date_str=None, pcd=None,
+                  sample_ids="'1'", **k):
+    """
+    holdout: pull # of returns in holdout period?
+    """
+    qname = "rec_freq_holdout" if holdout else "rec_freq"
+    kw = dict(
+        model_start_date_str=model_start_date_str,
+        sample_ids=sample_ids,
+        sample_comment='' if sample_ids else '--',
+        qname=qname,
+        pcd=pcd
+    )
+    kw.update(k)
+    kw = {k: v for k, v in kw.items() if v is not None}
+    return q.format(**kw)
+
+
+def run_rec_freq_spk(rfn_base_query=base_query, HO_WIN=14, MODEL_WIN=90,
+                     holdout=False, sample_ids="'1'",
+                     ho_start='2018-08-01', ho_days_in_future=None, spark=None):
+    """
+    holdout: whether to pull # of returns in holdout period. Useful
+        for evaluating model.
+    ho_days_in_future: int?; if 1, set the first day in holdout period
+        to be tomorrow. Negative numbers will set it to before today.
+    """
+    assert spark is not None, "Must pass SparkSession `spark`"
+    if ho_days_in_future is not None:
+        ho_start = dt.date.today() + dt.timedelta(days=ho_days_in_future)
+    r = mk_time_params(HO_WIN=HO_WIN, MODEL_WIN=MODEL_WIN, ho_start=ho_start)
+    r.q = mk_rec_freq_q(
+        q=rfn_base_query,
+        holdout=holdout,
+        # ignore_pcd=ignore_pcd,
+        sample_ids=sample_ids,
+        model_start_date_str=r.model_start_date_str, pcd=r.model_start_date,
+        ho_start_date=r.ho_start_date, ho_last_date_str=r.ho_last_date_str,
+        ho_start_date_str=r.ho_start_date_str
+    )
+#     print(r.q)
+    dfs = spark.sql(r.q)
+    return dfs, r.q
+
+
+def reduce_rec_freq_spk(dfs, frt_cols=['Frequency', 'Recency', 'T']):
+    """Reduce r/f/n spark dataframe to r/f/n pattern count.
+    This can be used for fitting.
+    """
+    df_red = dfs.groupby(frt_cols).count(
+    ).withColumnRenamed('count', 'n_users')
+    return df_red
+
+
+def rec_freq_spk2pandas(df_spk, MODEL_WIN):
+    df = df_spk.toPandas()
+    df = df.assign(Max_day=lambda x: pd.to_datetime(x.Max_day),
+                   Min_day=lambda x: pd.to_datetime(x.Min_day))
+    test_rft(df, duration=MODEL_WIN)
+    print('Recency/frequency Definition tests passed')
+    return df
+
+
+def run_rec_freq(HO_WIN=14, MODEL_WIN=90, sample_ids="'1'",
```

Comment:
This function doesn't expose the spark context to the `run_rec_freq_spk` function.

Reply:
appears you fixed this too